plugins/python/modulewrap.cpp

```diff
@@ -0,0 +1,632 @@
+    return nullptr;
+  }
+
+  PyObject* callable = PyTuple_GetItem(args, 0);
+  if (!PyCallable_Check(callable)) {
+    PyErr_SetString(PyExc_TypeError, "provided algorithm is not callable");
+    return nullptr;
+  }
+
+  // retrieve function name and argument types
+  PyObject* pyname = PyObject_GetAttrString(callable, "__name__");
+  if (!pyname) {
+    // AttributeError already set
+    return nullptr;
+  }
+  std::string cname = PyUnicode_AsUTF8(pyname);
+  Py_DECREF(pyname);
+
+  PyObject* inputs = PyTuple_GetItem(args, 1);
+  PyObject* outputs = PyTuple_GetItem(args, 2);
+  if (!(PySequence_Check(inputs) && PySequence_Check(outputs))) {
+    PyErr_SetString(PyExc_TypeError, "inputs and outputs need to be sequences");
+    return nullptr;
+  }
+
+  // convert input and output declarations, to be able to pass them to Phlex
+  auto const& cinputs = cseq(inputs);
+  auto const& coutputs = cseq(outputs);
+  if (coutputs.size() > 1) {
+    PyErr_SetString(PyExc_TypeError, "only a single output supported");
+    return nullptr;
+  }
+  std::string output = coutputs.empty() ? "" : coutputs[0];
+
+  std::vector<std::string> input_types;
+  input_types.reserve(cinputs.size());
+  std::string output_type; // TODO: accept a tuple (see also above)
+
+  PyObject* sann = PyUnicode_FromString("__annotations__");
+  PyObject* annot = PyObject_GetAttr(callable, sann);
+  if (!annot) {
+    // the callable may be an instance with a __call__ method
+    PyErr_Clear();
+    PyObject* callm = PyObject_GetAttrString(callable, "__call__");
+    if (callm) {
+      annot = PyObject_GetAttr(callm, sann);
+      Py_DECREF(callm);
+    }
+  }
+  Py_DECREF(sann);
+
+  if (annot && PyDict_Check(annot) && PyDict_Size(annot)) {
+    PyObject* ret = PyDict_GetItemString(annot, "return");
+    if (ret)
+      output_type = annotation_as_text(ret);
+
+    // dictionary is ordered with return last if provide (note: the keys here
+    // could be used as input labels, instead of the ones from the configuration,
+    // but that is probably not practical in actual use, so they are ignored)
+    PyObject* values = PyDict_Values(annot);
+    for (Py_ssize_t i = 0; i < (PyList_GET_SIZE(values) - (ret ? 1 : 0)); ++i) {
+      PyObject* item = PyList_GET_ITEM(values, i);
+      input_types.emplace_back(annotation_as_text(item));
+    }
+    Py_DECREF(values);
+  }
+  Py_XDECREF(annot);
+
+  if (input_types.size() != cinputs.size()) {
+    PyErr_SetString(PyExc_TypeError, "annotions not found or unknown formatting");
+    return nullptr;
+  }
+
+  // insert input converter nodes into the graph
+  for (size_t i = 0; i < (size_t)cinputs.size(); ++i) {
+    // TODO: this seems overly verbose and inefficient, but the function needs
+    // to be properly types, so every option is made explicit
+    auto const& inp = cinputs[i];
+    auto const& inp_type = input_types[i];
+
+    if (inp_type == "bool")
+      INSERT_INPUT_CONVERTER(bool, inp);
```

Comment:
If we have more than one Python algorithm using the same inputs, wouldn't this register the same conversion algorithm multiple times?